### ISSUE TITLE
questions using tritondse

### ISSUE THREAD

**Issue by vanhauser-thc** [quarkslab/tritondse#1]
original question removed, more below.

**Comment by RobinDavid** [comment]
Hello o/

Your script works. I put below a version slightly improved. Changes are:
* a added `pipe_stderr=True` in the config to output the bail printf result
* a changed the seed format to RAW which is the most simple form of input. (we use composite when the input is made of various chunks of bytes that need to be injected at scattered locations)

I guess you say it does not work because we don't cover the last one ? (Because it works great for me on "wrong string" 0, 4 and 12).

The issue (but it is not an issue) is the following. The condition `(len < 54)` checks that the len of the input string provided is less than 54 which is the case in all inputs provided and thus exit (we never reach the last if). The `len` is a *meta-property* of the input and the DSE by default cannot reason on it. As `len` here is not a symbolic variable there is nothing to negate for the symbolic engine.

We could have chosen to create a symbolic variable as the read result in order to reason symbolically on input length. However that raises multiple issue in the coherency of inputs that can be generated. That is a design choice for a DSE. We choose to let the user handling it by hand or providing an input large enough in the initial corpus not to have this issue. I think angr have special heuristics to handle such case.

For floating point instructions, yes Triton does not support them and as its the underlying engine we don't support them either.


```python
import sys
import logging

from tritondse import Config
from tritondse import CoverageStrategy
from tritondse import ProcessState
from tritondse import Program
from tritondse import Seed
from tritondse import SeedFormat
from tritondse import SymbolicExecutor
from tritondse import SymbolicExplorator


logging.basicConfig(level=logging.INFO)


def pre_exec_hook(se: SymbolicExecutor, state: ProcessState):
    logging.info(f"[PRE-EXEC] Processing seed: {se.seed.hash}, \
                    ({repr(se.seed.content)})")
    with open("out/" + se.seed.hash, 'wb') as file:
        file.write(se.seed.content)


# Load the program (LIEF-based program loader).
prog = Program(sys.argv[1])

# Load the configuration.
config = Config(coverage_strategy=CoverageStrategy.PATH,
               debug=True,
               pipe_stdout=True,
               pipe_stderr=True,
               seed_format=SeedFormat.RAW)

# Create an instance of the Symbolic Explorator
dse = SymbolicExplorator(config, prog)

# Create a starting seed, representing argv.
seed = Seed(b"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")

# Add seed to the worklist.
dse.add_input_seed(seed)

# Add callbacks.
dse.callback_manager.register_pre_execution_callback(pre_exec_hook)

# Start exploration!
dse.explore()
```

**Comment by vanhauser-thc** [comment]
yeah I wanted to use the raw seeds, but I found no example how to do that and didnt want to dig into the source :)

my statement was not very well phrased. yes it can partially solve that string compare example. But I do not understand why it is not able to solve it completely.

> We choose to let the user handling it by hand or providing an input large enough in the initial corpus not to have this issue.

I was sure I had it large enough but you are right that is the reason and then solves everything fine, also most other testcases like u128 :)

Is there a way to tell the script which expressions not to solve?
I ask that for fuzzing - usually you are interested only in new edges discovered and do not want to solve what was previously solved.

**Comment by RobinDavid** [comment]
If you use the `SymbolicExplorer` it will automatically **not** try solving what has already been solved. Let me explain further how it works. The explorer keeps a [`GlobalCoverage`](https://quarkslab.github.io/tritondse/api/coverage.html#tritondse.coverage.GlobalCoverage) object internally that keeps track of what has been covered and by extension what has not. Thus, after executing an input, the engine will try solving only those how have not been covered yet (according to the coverage strategy). You can have a look [here](https://github.com/quarkslab/tritondse/blob/main/tritondse/coverage.py#L347), the ``__iter_new_paths`` is a method of `GlobalCoverage` has this object knows what has been covered or not.

To save and restore that state you can simply call the [`to_file`](https://quarkslab.github.io/tritondse/api/coverage.html#tritondse.coverage.GlobalCoverage.to_file) and [`from_file`](https://quarkslab.github.io/tritondse/api/coverage.html#tritondse.coverage.GlobalCoverage.from_file).

For your use-case. I don't know if whenever you launch your script you want to execute exactly one input or potentially multiple, but by default the method `explore()` will not return until it has consumed recursively all the inputs in its queue. If you want to execute only one you might want to use `step()` combined with `post_exploration`.

**Comment by vanhauser-thc** [comment]
Thanks that helps me a lot :)
btw. I compared kirenenko, symcc, manticore, tritondse and fuzzolic now in https://github.com/AFLplusplus/fuzzer-challenges, and tritondse is the best solver currently.

I would have loved to test angr via driller/phuzzer, but I am unable to build and therefore test it.

maybe symsan and symcts will be better once their development is finished, but these are source-only solutions and require specific llvm versions making them less flexible.

so I will build a custom mutator for AFL++ with tritondse and it will be interesting to see how it improves on fuzzbench :)
(so there will be more questions coming from me, I am not a python fan, and I am not good at it ...)

**Comment by vanhauser-thc** [comment]
OK I got something mostly working as a test now:
https://github.com/AFLplusplus/AFLplusplus/tree/dev/custom_mutators/aflpp_tritondse

How can I suppress such warnings:
```
WARNING:root:calling getopt which is unsupported
```

Also I get this error although I carefully check that a seed is not supplied twice:
```
ERROR:root:seed not meant to be NEW at the end of execution (6009b27034ea1c25c4ed44db171c07fd) (dropped)
```

and maybe there are more thing I can enhance in the (so far still simple) script.
I appreciate your help!

**Comment by RobinDavid** [comment]
That's so nice ! I would love to see your experiment working and tritondse being used as a mutator 😍 .

Yes I would bet all engines relying on source-code would do better as they can leverage more information.

Ahah I am a Python fan so feel free to ask any questions 😉.

For the warning ```getopt``` it means we did not implemented a symbolic stub for that libc function. What happened in that case is that we just jump over the function and we don't apply side effects of that function. That's usually a bad news as it means we miss some side effects on the memory and the program might start behaving weirdly. You have two options for that:

* writing the symbolic stub for that function. That is a bit tedious. You have an example [here](https://github.com/quarkslab/tritondse/blob/main/tritondse/routines.py#L1059).
* Loading the program using the `CleLoader` object instead of `Program`. It will recursively load shared libraries of the program and use the function of the library if (no symbolic stub is present)

I'd definitely recommend using the 2nd.

The second issue is probably linked by the first. It means that the program execution ended without assigning the input a status (ok, crash). It assigns the status 'ok' when calling libc `exit()` or executing ``HLT`` instruction. I guess it did not terminated in a gentle manner.

Whenever I will have the time, I will try writing you a small snippet to receive an input, loading global coverage from file,  executing the input, generating new inputs and saving the global coverage back. We had not that scenario in mind so it might requires monkey patching object variables *(at the moment)*.

**Comment by vanhauser-thc** [comment]
> For the warning `getopt` it means we did not implemented a symbolic stub for that libc function. What happened in that case is that we just jump over the function and we don't apply side effects of that function. That's usually a bad news as it means we miss some side effects on the memory and the program might start behaving weirdly. You have two options for that:
> 
>     * writing the symbolic stub for that function. That is a bit tedious. You have an example [here](https://github.com/quarkslab/tritondse/blob/main/tritondse/routines.py#L1059).
> 
>     * Loading the program using the `CleLoader` object instead of `Program`. It will recursively load shared libraries of the program and use the function of the library if (no symbolic stub is present)

I cannot expect the user to implement symbolic stubs for every function in a program they want to fuzz tritondse doesnt know :-)
Do you have an example for how to use CleLoader so it does the same as Program?

> The second issue is probably linked by the first. It means that the program execution ended without assigning the input a status (ok, crash). It assigns the status 'ok' when calling libc `exit()` or executing `HLT` instruction. I guess it did not terminated in a gentle manner.

the program exits the program with a return of 0 or 1 from main.

> Whenever I will have the time, I will try writing you a small snippet to receive an input, loading global coverage from file, executing the input, generating new inputs and saving the global coverage back. We had not that scenario in mind so it might requires monkey patching object variables _(at the moment)_.

I do not need to save and restore the coverage,  I solved that. the python will be loaded into the program and therefore be persistent.
Primarily I need to solve the above two issues.

Then I need to find to get an idea how often to step() in the mutator so it makes sense ... maybe I measure time ... dunno.

**Comment by vanhauser-thc** [comment]
And is it ok that I add as many seeds with add_input_seed as I want even at later step()s?

**Comment by RobinDavid** [comment]
Yes, indeed we cannot except people to write symbolic stubs for all external function in a program. That's why we made the `CleLoader`. You can just is it as a droppin replacement of `Program`. They implement the same [`interface`](https://github.com/quarkslab/tritondse/blob/main/tritondse/loaders/loader.py#L33).

Please bear in mind that we don't implements syscalls, thus if the libc functions, or the program uses syscalls directly some side-effects will be missed and the emulation might behave weirdly. *(thats a problem of symbolic execution that must model everything to be correct)*

Ok great, it is better to keep the Python program persistent to avoid have this save/restore process. On a `SymbolicExecutor` object you can have how much time it [took to execute](https://quarkslab.github.io/tritondse/api/executor.html#tritondse.SymbolicExecutor.execution_time).

Yes you can definitely add as many seeds as you want with `add_input_seed` at any moment. You can also do it in callbacks that only takes the `SymbolicExecutor` as parameter. The method is [`enqueue_seed`](https://quarkslab.github.io/tritondse/api/executor.html#tritondse.SymbolicExecutor.enqueue_seed).

**Comment by vanhauser-thc** [comment]
I am using `CleLoader` now, but I still get
```
WARNING:root:symbol wmemchr imported but unsupported
WARNING:root:symbol strcspn imported but unsupported
WARNING:root:symbol wcscpy imported but unsupported
WARNING:root:symbol strcat imported but unsupported
```

**Comment by vanhauser-thc** [comment]
and a question about file inputs. lets say a target reads from /tmp/foo.txt.  I can create the file and write the input in pre_exec_hook. but additionally it will still be sent via stdin. how can I suppress that?

**Comment by RobinDavid** [comment]
For the functions 'imported but unsupported', I don't have straightforward explanation. Would mind sharing the input program you are working with ?

For the reading in /tmp/foo.txt, if you want it to become your input instead of (stdin/argv), you have to use the composite seeds. It can look like this:

```python
seeds = Seed(CompositeData(files={"stdin": b"",  # thus nothing on stdin
                                  "/tmp/foo.txt": b"AAAAAA"}))
```

When the file will be opened a pseudo file descriptor will be assigned. And then when a `fread` or `read` happen we check if the file descriptor is an input one, and if so we inject the input in the resulting buffer [(code)](https://github.com/quarkslab/tritondse/blob/main/tritondse/routines.py#L753).
Thus if you are in a regular case, you don't need to write the file in a pre_exec_hook.

If you have a "descriptor not found", it means the file has not been previously opened via a `fopen` etc, and that we don't find it in the [`fd_table`](https://github.com/quarkslab/tritondse/blob/main/tritondse/process_state.py#L79) of a `ProcessState`object. Can you pinpoint me the code snippet that does that ?

**Comment by RobinDavid** [comment]
Sorry for the late reply.

I looked at it and what happen is fairly simple. We don't have a symbolic stub for `open`, thus thanks to the `CleLoader` we call the libc `open` function`. However this one ultimately perform an `openat` syscall. As we don't support syscalls, we simply jump over the syscall and end up with an invalid file descriptor.

I first show I solved it without touching the `tritondse` internal code. I first needed to implement a function stub for open (inspired by the [`fopen`](https://github.com/quarkslab/tritondse/blob/main/tritondse/routines.py#L563) we already have)

```python
def rtn_open(se: SymbolicExecutor, pstate: ProcessState, pc):
    """
    The open behavior.
    """
    logging.debug('open hooked')

    # Get arguments
    arg0 = pstate.get_argument_value(0)  # const char *pathname
    flags = pstate.get_argument_value(1)  # int flags
    mode = pstate.get_argument_value(2)  # int mode
    arg0s = pstate.memory.read_string(arg0)

    # Concretize the whole path name
    pstate.concretize_memory_bytes(arg0, len(arg0s)+1)  # Concretize the whole string + \0

    # We use flags as concrete value
    pstate.concretize_argument(1)

    # Use the flags to open the file in the write mode.
    mode = ""
    if (flags & 0xFF) == 0x00:   # O_RDONLY
        mode = "r"
    elif (flags & 0xFF) == 0x01: # O_WRONLY
        mode = "w"
    elif (flags & 0xFF) == 0x02: # O_RDWR
        mode = "r+"

    if (flags & 0x0100): # O_CREAT
        mode += "x"
    if (flags & 0x0200): # O_APPEND
        mode = "a"  # replace completely value

    if se.seed.is_file_defined(arg0s) and "r" in mode:  # input file and opened in reading
        logging.info(f"opening an input file: {arg0s}")
        # Program is opening an input
        data = se.seed.get_file_input(arg0s)
        filedesc = pstate.create_file_descriptor(arg0s, io.BytesIO(data))
        fd = filedesc.id
    else:
        # Try to open it as a regular file
        try:
            fd = open(arg0s, mode)  # use the mode here
            filedesc = pstate.create_file_descriptor(arg0s, fd)
            fd = filedesc.id
        except Exception as e:
            logging.debug(f"Failed to open {arg0s} {e}")
            fd = pstate.minus_one

    pstate.write_register("rax", fd)  # write the return value
    pstate.cpu.program_counter = pstate.pop_stack_value()  # pop the return value
    se.skip_instruction()  # skip the current instruction so that the engine go straight fetching the next instruction
```

Then you can set a hook on that function when its being called so that this python snippet gets called instead:

```python
dse.callback_manager.register_function_callback("open", rtn_open)
```

**Comment by RobinDavid** [comment]
To see what happened I used a simple built-in probes that show the execution trace:

```python
from tritondse.probes.basic_trace import BasicDebugTrace

[snip]

dse.callback_manager.register_probe(BasicDebugTrace())
```
It prints all instructions executed, that's how I have seen that we entered the open and that it did a syscall.

**Comment by RobinDavid** [comment]
Still, I agree its a shame we don't have this built-in, and it would be nice to have warning when executing a syscall *(which if not implemented can drive the symbolic execution in unexpected states)*.

I did a new release: https://github.com/quarkslab/tritondse/releases/tag/v0.1.3  (available in Pypi).

Thus it should work out-of-the box without changes on your side 😉 .

**Comment by vanhauser-thc** [comment]
thanks so far!
btw I tried your hook approach but it does not work:
```
WARNING:root:can't find function open in <tritondse.loaders.cle_loader.CleLoader object at 0x7fffee98ecd0>
WARNING:root:File descriptor (257) not found
```
super weird.

**Comment by vanhauser-thc** [comment]
gdb shows in main:
```
   0x0000000000003549 <+377>:	mov    rdi,QWORD PTR [rbx+0x10]
   0x000000000000354d <+381>:	xor    esi,esi
   0x000000000000354f <+383>:	xor    eax,eax
   0x0000000000003551 <+385>:	call   0x3240 <open@plt>
```

**Comment by RobinDavid** [comment]
You have the message with the **v0.1.3** version ?

I fixed the function ``find_function_addr`` of the CleLoader [here](https://github.com/quarkslab/tritondse/commit/1b9a091028c94bbaa3d77fae06871bc757de804a#diff-606272635d77b7f0610b161e01e3d5b4af41d0a45d565c210f28df9035e60f16L178).

It was using the function from the main class which does nothing. That really was a bug..

**Comment by vanhauser-thc** [comment]
I just upgraded to 0.1.3 and there it works, also the `WARNING:root:symbol strnlen imported but unsupported` warnings are now gone :)
so it looks like everything is working for me now, thank you so far!

**Comment by vanhauser-thc** [comment]
hmm but now I get an exception in tritondse:
```
Exception in thread [exec:00000000]:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.11/dist-packages/tritondse/symbolic_explorator.py", line 146, in _worker
    execution.run(self._executor_stop_at)
  File "/usr/local/lib/python3.11/dist-packages/tritondse/symbolic_executor.py", line 578, in run
    self.__emulate()
  File "/usr/local/lib/python3.11/dist-packages/tritondse/symbolic_executor.py", line 379, in __emulate
    self._routines_handler(instruction)
  File "/usr/local/lib/python3.11/dist-packages/tritondse/symbolic_executor.py", line 430, in _routines_handler
    ret_val = routine(self, self.pstate)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/tritondse/routines.py", line 851, in rtn_write
    fdesc.fd.write(data)
TypeError: write() argument must be str, not bytes
```
I fixed that in routines.py with
```
            fdesc.fd.write(str(data))
```

**Comment by RobinDavid** [comment]
Oh damn. Maybe I went too fast, I need to add 'b' in the mode when opening the file.
As we require Seed inputs to be bytes, we only want have files opened in 'bytes' mode (not to interleave str / bytes). Otherwise its a mess to handle.

The problem here is that you probably open the file in text mode, but you intent to write python data that are bytes, thus it conflicts. In tritondse we want the python `open` to be always with bytes to avoid this issue.

I am going to change that.

**Comment by vanhauser-thc** [comment]
btw there are still prints like

```
WARNING:root:execute syscall instruction 262
WARNING:root:execute syscall instruction 9
ERROR:root:Instruction not mapped: 0x26156
```

is there a way to suppress these? they are messing with the UI :)
Having these when debugging is valuable, but for productive runs with is annoying as it gets printed with every dse.step()

**Comment by RobinDavid** [comment]
No unfortunately I am writing directly in the root logger. If you disable logging they will disappear but I guess you also want to do some logging on your side.

The syscall warning are the one I added so that its clear for user when they are executing syscalls (and thus might have weird behavior). The 'Instruction not mapped', should be investigated. It means your program counter (rip or pc) is fetching an instruction in a memory area not mapped. So if its the triggering of a bug that's okay, but otherwise it means your execution is going nuts.

I'll write logging in a dedicated logger, in a later release.

**Comment by RobinDavid** [comment]
I created an issue for that: https://github.com/quarkslab/tritondse/issues/4

**Comment by vanhauser-thc** [comment]
ok I am done implementing the custom mutator for tritondse. sadly I cannot test it on fuzzbench as I get weird errors with python, it cannot build hashlib on a freshly installed python3.11 from source etc.

do you know of anything in tritondse that could be improved on to make it better?
I tried it against symqemu with the same input and it solved edges that symqemu did not but the same the other way around :)

**Comment by RobinDavid** [comment]
Nice for the custom mutator. That's unfortunate. Can you pinpoint us what you have done so that we can try replaying the errors ? @cnheitman worked on pastis fuzzbench integration he might have some ideas.

You mean improved so that it can be integrated in fuzzbench or in general ?

**Comment by vanhauser-thc** [comment]
> Nice for the custom mutator. That's unfortunate. Can you pinpoint us what you have done so that we can try replaying the errors ? @cnheitman worked on pastis fuzzbench integration he might have some ideas.

true I could copy the python setup from pastis, would be much simpler than me trying for hours to get things work ... oh my ...

> You mean improved so that it can be integrated in fuzzbench or in general ?

improved as more/better/faster solving, more libcall/syscalls to support etc.

**Comment by vanhauser-thc** [comment]
@RobinDavid 
I now get the following tritondse error:
```
  File "/prg/afl++/custom_mutators/aflpp_tritondse/aflpp_tritondse.py", line 6, in <module>
    from tritondse import CleLoader
  File "/usr/local/lib/python3.11/dist-packages/tritondse/__init__.py", line 1, in <module>
    from .config                import Config
  File "/usr/local/lib/python3.11/dist-packages/tritondse/config.py", line 9, in <module>
    from tritondse.coverage import CoverageStrategy, BranchSolvingStrategy
  File "/usr/local/lib/python3.11/dist-packages/tritondse/coverage.py", line 15, in <module>
    from triton import AST_NODE
ImportError: cannot import name 'AST_NODE' from 'triton' (/usr/local/lib/python3.11/dist-packages/triton/__init__.py)
```

**Comment by RobinDavid** [comment]
Hi @vanhauser-thc!

This is a Triton issue. The path */usr/local/lib/python3.11/dist-packages/triton/__init__.py* is weird. How did you installed Triton ? Have you installed it through pip ? (I would say no).
My guess is that when you pip-installed tritondse, the pip module `triton-library` got correctly installed, but it is another version that gets imported (probably from a previous installation).

On my side the triton module is a shared library when importing AST_NODE:

```python
import sys, triton
print(sys.modules['triton'])
#<module 'triton' from '/home/robin/[...]/site-packages/triton.cpython-311-x86_64-linux-gnu.so'>
```
